### PR TITLE
Add entries to styleguide

### PR DESCRIPTION
**Pull request by renancaraujo**
# Description

This proposes some additions to our style guide. We can add more and use [this](https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo) as a reference.

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] My PR includes unit or integration tests for *all* changed/updated/fixed behaviors and are passing (See [Contributor Guide]).
- [x] I updated/added relevant documentation (doc comments with `///`) and updated/added examples in `doc/examples`.
- [x] I have formatted my code with `./scripts/format.sh` and the Flame analyzer (`./scripts/analyze.sh`) does not report any problems.
- [x] I read and followed the [Flame Style Guide].
- [x] I have added a description of the change under `[next]` in `CHANGELOG.md`.
- [ ] I removed the `Draft` status, by clicking on the `Ready for review` button in this PR.
- [x] I am willing to follow up on review comments in a timely manner.

## Breaking Change

Does your PR require Flame users to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change (please indicate a breaking change in `CHANGELOG.md`).
- [ ] No, this is *not* a breaking change.

## Related Issues

[issue database]: https://github.com/flame-engine/flame/issues
[Contributor Guide]: https://github.com/flame-engine/flame/blob/main/CONTRIBUTING.md
[Flame Style Guide]: https://github.com/flame-engine/flame/blob/main/STYLEGUIDE.md

### CODE REVIEW

File: STYLEGUIDE.md [hunk 1]
```diff
@@ -1,8 +1,14 @@
 # Style Guide
 
-This is WIP! It is definitely not complete. We just starting putting this together.
+This is a general style guide that shall govern over the Flame Engine repository. The aim is to keep a common stable general environment.
```
Comment (spydon) [inline STYLEGUIDE.md:3]: Max 100 in line length

File: STYLEGUIDE.md [hunk 2]
```diff
@@ -1,8 +1,14 @@
...
+This includes high level guidance to help on simple decisions in the day-to-day development life.
 
-## Trailing Commas and Wrapping
+This extends rules on the official [Flutter Style Guide](https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo).
+
```
Comment (spydon) [inline STYLEGUIDE.md:7]:
```suggestion
```

File: STYLEGUIDE.md [hunk 3]
```diff
@@ -1,8 +1,14 @@
 # Style Guide
 
-This is WIP! It is definitely not complete. We just starting putting this together.
+This is a general style guide that shall govern over the Flame Engine repository. The aim is to keep a common stable general environment.
+This includes high level guidance to help on simple decisions in the day-to-day development life.
```
Comment (spydon) [inline STYLEGUIDE.md:4]:
```suggestion
This includes high level guidance to help with simple decisions in the day-to-day development life.
```

File: STYLEGUIDE.md [hunk 4]
```diff
@@ -72,3 +78,74 @@ import 'camera.dart';
 import 'game.dart';
 import 'viewport.dart';
 ```
+
+### Identifiers
+
+Use all [effective dart patterns](https://dart.dev/guides/language/effective-dart)
+
+* Avoid using global constants, but if you do use, name it prefixing with a `k`.
```
Comment (spydon) [inline STYLEGUIDE.md:86]: Interesting, never knew about the `k` thing

File: STYLEGUIDE.md [hunk 5]
```diff
@@ -72,3 +78,74 @@ import 'camera.dart';
...
+const someConstant = 42;
+
+// Do instead:
+const kSomeConstant = 42;
+```
+
+### Comments
+
+* Add an identifier for TODO comments
+* Use full capitalized TODO
```
Comment (spydon) [inline STYLEGUIDE.md:99]:
```suggestion
* Use fully capitalized TODO notation
```

File: STYLEGUIDE.md [hunk 6]
```diff
@@ -72,3 +78,74 @@ import 'camera.dart';
...
+
+```dart
+
+// Do not:
+// Todo: This thing should be that thing
+final thisThing = 13;
+
+// Do instead:
+// TODO(wofstein): This thing should be that thing
```
Comment (spydon) [inline STYLEGUIDE.md:108]:
```suggestion
// TODO(wolfenrain): This thing should be that thing
```
:laughing:

File: STYLEGUIDE.md [hunk 7]
```diff
@@ -72,3 +78,74 @@ import 'camera.dart';
...
+const thisThing = 13;
+```
+
+### Asserts
+
+* Use asserts to detect contract violation.
+* Use asserts to run debug only code
```
Comment (spydon) [inline STYLEGUIDE.md:115]:
```suggestion
* Use asserts to run debug only code.
```
Seems to be full stop in the items in the other lists.

File: STYLEGUIDE.md [hunk 8]
```diff
@@ -72,3 +78,74 @@ import 'camera.dart';
...
+
+
+
+## Code documentation 
+
+Specific guides for dartdoc documentation.
+
+#### Use impersonal tone
+
+Avoid using words such as "you", "ours" and "yours" on code documentation
```
Comment (spydon) [inline STYLEGUIDE.md:138]: I think we need to review the docs if we want to enforce this rule. Might be quick though.

**Review by spydon — COMMENTED** [review]
Looking good, just a few comments

File: STYLEGUIDE.md (hunk 1 above)
Comment (spydon) [reply STYLEGUIDE.md:3]: Oh, this should also be mentioned in this doc I guess

File: STYLEGUIDE.md (hunk 6 above)
Comment (renancaraujo) [reply STYLEGUIDE.md:107]: I reffuse.

File: STYLEGUIDE.md (hunk 4 above)
Comment (renancaraujo) [reply STYLEGUIDE.md:85]: Fun fact: thats the origin of kPop.

File: STYLEGUIDE.md (hunk 6 above)
Comment (wolfenrain) [reply STYLEGUIDE.md]: If anything it should have been `wolfstein`.

**Review by wolfenrain — APPROVED** [review]

File: STYLEGUIDE.md [hunk 9]
```diff
@@ -37,7 +41,7 @@ foo(f1,
     f2)
 ```
 
-## Imports
+### Imports
```
Comment (luanpotter) [inline STYLEGUIDE.md:44]: @renancaraujo you added a lot of things but didn't fix the one TODO we had:

```
TODO Do/Do not use the imports grouping internally (like component.dart, extensions.dart, etc).
```

File: STYLEGUIDE.md (hunk 4 above)
Comment (luanpotter) [reply STYLEGUIDE.md:84]: I don't quite agree with this rule, I think we should either ban it altogether or allow. prefixing things with k is weird

File: STYLEGUIDE.md [hunk 10]
```diff
@@ -72,3 +76,74 @@ import 'camera.dart';
...
+const someConstant = 42;
+
+// Do instead:
+const kSomeConstant = 42;
+```
+
+### Comments
+
+* Add an identifier for TODO comments
+* Use fully capitalized TODO notation
```
Comment (luanpotter) [inline STYLEGUIDE.md:97]: I think we can be more clear adding a "template":

```
// TODO(username): Comment
```

File: STYLEGUIDE.md [hunk 11]
```diff
@@ -72,3 +76,74 @@ import 'camera.dart';
...
+
+// Do instead:
+// TODO(wolfenrain): This thing should be that thing
+const thisThing = 13;
+```
+
+### Asserts
+
+* Use asserts to detect contract violation.
+* Use asserts to run debug only code.
```
Comment (luanpotter) [inline STYLEGUIDE.md:113]: also don't quite like this, seems hacky. I think we can leave it out until we ever need something like this and then we can discuss.

File: STYLEGUIDE.md [hunk 12]
```diff
@@ -72,3 +76,74 @@ import 'camera.dart';
...
+Some elements may benefit from a simple usage example.
+
+#### Avoid useless code documentation
+
+Avoid documentation that just repeats the obvious. 
+For example, `void doStuff()` be documented as "Method that does stuff".
+
+#### Consider adding linkage between docs
+
+Consider adding a "See also" section to element documentation
```
Comment (luanpotter) [inline STYLEGUIDE.md:149]: also make sure to mark code elements with `[]`

**Review by luanpotter — COMMENTED** [review]
Some comments, most importantly, we should fix the TODO!

File: STYLEGUIDE.md (hunk 4 above)
Comment (erickzanardo) [reply STYLEGUIDE.md:84]: @luanpotter the `k` thing is how Flutter does on its source base, do I think this rule is fine.

File: STYLEGUIDE.md [hunk 13]
```diff
@@ -72,3 +76,74 @@ import 'camera.dart';
...
+  assert((){
+    // debug only code
+  });
+  
+  //...
+}
+
+````
+
+
```
Comment (erickzanardo) [inline STYLEGUIDE.md:128]:
```suggestion
```

File: STYLEGUIDE.md [hunk 14]
```diff
@@ -72,3 +76,74 @@ import 'camera.dart';
...
+````dart
+void something(int smaller, int bigger) {
+  assert(small < bigger, "smaller is not smaller than bigger");
+  assert((){
+    // debug only code
+  });
+  
+  //...
+}
+
```
Comment (erickzanardo) [inline STYLEGUIDE.md:125]:
```suggestion
```

**Review by erickzanardo — APPROVED** [review]
Few nits, but LGTM already

File: STYLEGUIDE.md (hunk 12 above)
Comment (renancaraujo) [reply STYLEGUIDE.md:149]: Well, this is basic dartdoc, do we really include this here?

File: STYLEGUIDE.md (hunk 11 above)
Comment (renancaraujo) [reply STYLEGUIDE.md:113]: done

File: STYLEGUIDE.md (hunk 10 above)
Comment (renancaraujo) [reply STYLEGUIDE.md:97]: done

File: STYLEGUIDE.md (hunk 9 above)
Comment (renancaraujo) [reply STYLEGUIDE.md:44]: done

**Review by spydon — APPROVED** [review]
Looks good, just the line length 100 missing

File: STYLEGUIDE.md (hunk 1 above)
Comment (renancaraujo) [reply STYLEGUIDE.md:3]: done

File: STYLEGUIDE.md (hunk 8 above)
Comment (renancaraujo) [reply STYLEGUIDE.md:136]: 151 occurrences just for `you` and `yours`. I think I will do this in a  separate PR.

File: STYLEGUIDE.md (hunk 4 above)
Comment (luanpotter) [reply STYLEGUIDE.md:93]: yeah but I think Flutter has reasons to need global variables, I don't think audioplayers should ever pollute the global scope like that.

Comment (luanpotter) [reply STYLEGUIDE.md:93]: I mean flame haha

File: STYLEGUIDE.md (hunk 12 above)
Comment (luanpotter) [reply STYLEGUIDE.md:152]: people might not know about it, I don't see why not make our preferred styles explicit

File: STYLEGUIDE.md (hunk 4 above)
Comment (luanpotter) [reply STYLEGUIDE.md:93]: if we ever want to add a global constant for some reason we can revisit this rule, I see no reason to encourage this behaviour now

File: STYLEGUIDE.md (hunk 8 above)
Comment (luanpotter) [reply STYLEGUIDE.md:139]: I would omit the rule for now; in a followup we can add the rule + fix and/or discuss if we need to change everything

Comment (luanpotter) [reply STYLEGUIDE.md:139]: Moved to an issue so we can discuss further and separately: https://github.com/flame-engine/flame/issues/767

**Review by spydon — APPROVED** [review]
Lgtm!

**Review by erickzanardo — APPROVED** [review]
LGTM

**Pull request merged**